File: lib/src/test/java/io/xj/lib/music/AccidentalTest.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.
package io.xj.lib.music;

import org.junit.Test;

import static io.xj.lib.music.Accidental.Flat;
import static io.xj.lib.music.Accidental.None;
import static io.xj.lib.music.Accidental.Sharp;
import static org.junit.Assert.assertEquals;

public class AccidentalTest {

  @Test
  public void TestAdjSymbolOf() {
    assertEquals(Sharp, Accidental.of("C"));
    assertEquals(Flat, Accidental.of("CMb5b7"));
    assertEquals(Sharp, Accidental.of("C#"));
    assertEquals(Flat, Accidental.of("Gb"));
    assertEquals(Flat, Accidental.of("G♭M"));
    assertEquals(Sharp, Accidental.of("A#m"));
    assertEquals(Sharp, Accidental.of("A♯M♯5"));
    assertEquals(Flat, Accidental.of("C minor"));
    assertEquals(Flat, Accidental.of("C dim"));
    assertEquals(Sharp, Accidental.of("CM M9 m7")); // More Sharpish than Flattish
    assertEquals(Flat, Accidental.of("Cm m9 M7"));  // More Flattish than Sharpish
    assertEquals(Sharp, Accidental.of("C major"));
  }

  @Test
  public void TestAdjSymbolBegin() {
    assertEquals(None, Accidental.firstOf("C".substring(1)));
    assertEquals(None, Accidental.firstOf("CMb5b7".substring(1)));
    assertEquals(Sharp, Accidental.firstOf("C#".substring(1)));
    assertEquals(Flat, Accidental.firstOf("Gb".substring(1)));
    assertEquals(Flat, Accidental.firstOf("G♭M".substring(1)));
    assertEquals(Sharp, Accidental.firstOf("A#m".substring(1)));
    assertEquals(Sharp, Accidental.firstOf("A♯M♯5".substring(1)));
  }


  @Test
  public void replaceWithExplicit() {
    assertEquals("Cs", Accidental.replaceWithExplicit("C#"));
    assertEquals("Cs", Accidental.replaceWithExplicit("C♯"));
    assertEquals("Cb", Accidental.replaceWithExplicit("Cb"));
    assertEquals("Cb", Accidental.replaceWithExplicit("C♭"));
  }
}

File: engine/test/music/AccidentalTest.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <gtest/gtest.h>

#include "xjmusic/music/Accidental.h"

using namespace XJ;


TEST(Music_Accidental, TestAdjSymbolOf) {
  ASSERT_EQ(Accidental::Sharp, accidentalOf("C"));
  ASSERT_EQ(Accidental::Flat, accidentalOf("CMb5b7"));
  ASSERT_EQ(Accidental::Sharp, accidentalOf("C#"));
  ASSERT_EQ(Accidental::Flat, accidentalOf("Gb"));
  ASSERT_EQ(Accidental::Flat, accidentalOf("GbM"));
  ASSERT_EQ(Accidental::Sharp, accidentalOf("A#m"));
  ASSERT_EQ(Accidental::Sharp, accidentalOf("A#M#5"));
  ASSERT_EQ(Accidental::Flat, accidentalOf("C minor"));
  ASSERT_EQ(Accidental::Flat, accidentalOf("C dim"));
  ASSERT_EQ(Accidental::Sharp, accidentalOf("CM M9 m7")); // More Sharpish than Flattish
  ASSERT_EQ(Accidental::Flat, accidentalOf("Cm m9 M7"));  // More Flattish than Sharpish
  ASSERT_EQ(Accidental::Sharp, accidentalOf("C major"));
}


TEST(Music_Accidental, TestAdjSymbolBegin) {
  ASSERT_EQ(Accidental::Natural, accidentalOfBeginning(""));
  ASSERT_EQ(Accidental::Natural, accidentalOfBeginning("Mb5b7"));
  ASSERT_EQ(Accidental::Sharp, accidentalOfBeginning("#"));
  ASSERT_EQ(Accidental::Flat, accidentalOfBeginning("b"));
  ASSERT_EQ(Accidental::Flat, accidentalOfBeginning("bM"));
  ASSERT_EQ(Accidental::Sharp, accidentalOfBeginning("#m"));
  ASSERT_EQ(Accidental::Sharp, accidentalOfBeginning("#M#5"));
}
